Handle migrated versions metadata table auto-update

goetas opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | #788 #828

#### Summary

This PR:
- adds a command `migrations:sync-metadata-storage` that ensure the table scheme `doctrine_migrations` is always up to date
- auto-migrates the v2 table structure to the v3 table structure
- each command before starting checks if the metadata table is up to date


~The `Allow phpunit 7.5` commit should be removed when https://github.com/doctrine/migrations/pull/879 will be merged~

goetas commented:
@jwage From the git history, in 2.x you have implemented the https://github.com/doctrine/migrations/tree/2.2.x/lib/Doctrine/Migrations/Tracking feature, that was supposed to do something similar as this PR. Do you have time to review this please?

greg0ire commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -37,46 +43,25 @@ final class TableMetadataStorage implements MetadataStorage
+    /** @var MigrationRepository|null */
+    private $migrationRepository;
+
+    public function __construct(Connection $connection, ?MetadataStorageConfiguration $configuration = null, ?MigrationRepository $migrationRepository = null)
```
```suggestion
    public function __construct(
        Connection $connection,
        ?MetadataStorageConfiguration $configuration = null,
        ?MigrationRepository $migrationRepository = null
    )
```

greg0ire commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -138,4 +123,96 @@ public function complete(ExecutionResult $result) : void
+    {
+        $schemaChangelog = new Table($this->configuration->getTableName());
+
+        $schemaChangelog->addColumn($this->configuration->getVersionColumnName(), 'string', ['notnull' => true, 'length' => $this->configuration->getVersionColumnLength()]);
```
```suggestion
        $schemaChangelog->addColumn(
            $this->configuration->getVersionColumnName(), 
            'string',
            ['notnull' => true, 'length' => $this->configuration->getVersionColumnLength()]
        );
```

greg0ire commented on lib/Doctrine/Migrations/Metadata/Storage/TableMetadataStorage.php:
```diff
@@ -138,4 +123,96 @@ public function complete(ExecutionResult $result) : void
+
+        foreach ($availableMigrations as $availableMigration) {
+            foreach ($executedMigrations as $k => $executedMigration) {
+                if (strpos((string) $availableMigration->getVersion(), (string) $executedMigration->getVersion()) !== (strlen((string) $availableMigration->getVersion()) - strlen((string) $executedMigration->getVersion()))) {
```
```suggestion
                if (strpos(
                    (string) $availableMigration->getVersion(),
                    (string) $executedMigration->getVersion()
                ) !== (
                    strlen((string) $availableMigration->getVersion()) - 
                    strlen((string) $executedMigration->getVersion())
                )) {
```

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/SyncMetadataCommand.php:
```diff
@@ -0,0 +1,40 @@
+
+        $this
+            ->setAliases(['sync-metadata-storage'])
+            ->setDescription('Ensures that the metadata storage is to the latest version.')
```
```suggestion
            ->setDescription('Ensures that the metadata storage is at the latest version.')
```

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/Command/SyncMetadataCommand.php:
```diff
@@ -0,0 +1,40 @@
+    public function execute(
+        InputInterface $input,
+        OutputInterface $output
+    ) : ?int {
```
```suggestion
    ) : int {
```

greg0ire commented:
I think something went wrong with your last push, can you please check? None of my comments are addressed now…

goetas commented:
@greg0ire my bad
- I forgot to pull the changes done via github UI. Now i have addressed your suggestions again. 
- I have also removed the temporary commit 
- i have improved the condition in https://github.com/doctrine/migrations/pull/878#discussion_r352330241 by extracting it in a private method, now is much more clear the propose of that if condition

goetas commented:
Since this feature is quite critical, can anybody else review/test it please?

alcaeus commented:
@goetas I don't have any projects using migrations, and other people might be wondering this too: what would be the best way to test this?

goetas commented:
The best way to test this is to have a project using migrations 2.x, try the upgrade process as described in https://github.com/doctrine/migrations/blob/master/UPGRADE.md, and share the results.
